src/Helpers/JWKFetcher.php

```diff
@@ -38,43 +45,69 @@ public function __construct(CacheHandler $cache = null, $guzzleOptions = [])
     }
 
     /**
+     * Convert a certificate to PEM.
+     *
+     * @param string $cert Certificate to convert.
      *
-     * @param  string $cert
      * @return string
      */
     protected function convertCertToPem($cert)
     {
-        return '-----BEGIN CERTIFICATE-----'.PHP_EOL.chunk_split($cert, 64, PHP_EOL).'-----END CERTIFICATE-----'.PHP_EOL;
+        $output  = '-----BEGIN CERTIFICATE-----'.PHP_EOL;
+        $output .= chunk_split($cert, 64, PHP_EOL);
+        $output .= '-----END CERTIFICATE-----'.PHP_EOL;
+        return $output;
     }
 
     /**
+     * Fetch JWKS from the token issuer.
      *
-     * @param string $iss
+     * @param string      $domain    Typically the issuer for the JWKS being fetched; include a trailing slash.
+     * @param null|string $jwks_path Path to the JWKS from the issuer domain; pass null to use $domain as the URL.
      *
      * @return array|mixed|null
-     *
-     * @throws \Exception
      */
-    public function fetchKeys($iss)
+    public function fetchKeys($domain, $jwks_path = '.well-known/jwks.json')
     {
-        $url = "{$iss}.well-known/jwks.json";
+        // If the path parameter is empty, use the $domain parameter as a complete URL.
+        if (empty( $jwks_path )) {
+            $domain_parsed = parse_url( $domain );
+            $domain        = $domain_parsed['scheme'].'://'.$domain_parsed['host'];
+            $domain       .= empty( $domain_parsed['port'] ) ? '' : ':'.$domain_parsed['port'];
+            $domain       .= '/';
+            $jwks_path     = empty( $domain_parsed['path'] ) ? '' : substr( $domain_parsed['path'], 1 );
+        }
+
+        // Check if we have a cache entry for this URL.
+        $url    = $domain.$jwks_path;
+        $secret = $this->cache->get($url);
+        if (! is_null($secret)) {
+            return $secret;
+        }
+
+        $secret = [];
 
-        if (($secret = $this->cache->get($url)) === null) {
-            $secret = [];
+        // GET the JWKS URL.
+        $request = new RequestBuilder([
+            'domain' => $domain,
+            'basePath' => $jwks_path,
+            'method' => 'GET',
+            'guzzleOptions' => $this->guzzleOptions
+        ]);
+        $jwks    = $request->call();
 
-            $request = new RequestBuilder([
-                'domain' => $iss,
-                'basePath' => '.well-known/jwks.json',
-                'method' => 'GET',
-                'guzzleOptions' => $this->guzzleOptions
-            ]);
-            $jwks    = $request->call();
+        // Nothing to return.
+        if (empty( $jwks['keys'] ) || ! is_array( $jwks['keys'] )) {
```

Comment:
wouldn't this explode if `jwks` is undefined? (failed request scenario)

Reply:
Failed request = thrown exception. Also, if `$jwks` is `null`, then `empty($jwks['key'])` will be `true`;